python/rmm/rmm.py

```diff
@@ -193,7 +194,8 @@ def rmm_cupy_allocator(nbytes):
     if cupy is None:
         raise ModuleNotFoundError("No module named 'cupy'")
 
-    buf = librmm.device_buffer.DeviceBuffer(size=nbytes)
+    stream = Stream(obj=cupy.cuda.get_current_stream())
+    buf = librmm.device_buffer.DeviceBuffer(size=nbytes, stream=stream)
```

Comment:
Is the CuPy stream guaranteed to not be destroyed before this DeviceBuffer? Or alternatively, does the user have sufficient control to prevent it from being destroyed before the DeviceBuffer?  If not, then you could get a CUDA error on destruction of the DeviceBuffer, or possibly worse, a very subtle race condition due to the CUDA runtime recycling stream handles. (Something we are just learning about.)

Reply:
We should probably hold a reference to the `Stream` object in the `DeviceBuffer` object

Reply:
I think it's a StreamView, isn't it? If so, holding a reference won't prevent the underlying `cudaStream_t` from being destroyed.

Reply:
`Stream` in Python holds a reference to an owning object `CudaStream`, Cupy Stream, Numba Stream, etc.

Reply:
But if the stream was created outside of RMM there's no way it can be owned by RMM.

Reply:
> But if the stream was created outside of RMM there's no way it can be owned by RMM.

It doesn't need to be. The RMM Python library `Stream` class will hold a reference to a Python object which is what keeps the underlying `cudaStream_t` from being freed. We just need to keep a reference to a `Stream` instance to keep things alive.

Reply:
This is a good point, I'll work on adding an attribute to hold a reference to the original CuPy stream.

Reply:
I did a bit of work on this, adding a new attribute `Stream _stream` to `DeviceStream` that gets populated with a reference to the original stream:

```diff
diff --git a/python/rmm/_lib/device_buffer.pxd b/python/rmm/_lib/device_buffer.pxd
index 98a79d1..cad62b4 100644
--- a/python/rmm/_lib/device_buffer.pxd
+++ b/python/rmm/_lib/device_buffer.pxd
@@ -37,6 +37,7 @@ cdef extern from "rmm/device_buffer.hpp" namespace "rmm" nogil:

 cdef class DeviceBuffer:
     cdef unique_ptr[device_buffer] c_obj
+    cdef Stream _stream

     @staticmethod
     cdef DeviceBuffer c_from_unique_ptr(unique_ptr[device_buffer] ptr)
diff --git a/python/rmm/_lib/device_buffer.pyx b/python/rmm/_lib/device_buffer.pyx
index b29bff1..f09957c 100644
--- a/python/rmm/_lib/device_buffer.pyx
+++ b/python/rmm/_lib/device_buffer.pyx
@@ -68,6 +68,8 @@ cdef class DeviceBuffer:
         cdef const void* c_ptr
         cdef cudaError_t err

+        self._stream = stream
+
         with nogil:
             c_ptr = <const void*>ptr
```

This works well, I also added a CuPy test that segfaults and with the above it doesn't. However, this increases the size of the `DeviceBuffer` object, which is tested against the data size that's allocated in https://github.com/rapidsai/rmm/blob/21e565065fbb85c450c2fa87fe4bc0c9ffe9ba63/python/rmm/tests/test_rmm.py#L112 , and that test now fails. My question here is, should I adjust the test to reflect the changes above or is there a reason which I'm not aware of where `DeviceBuffer` is expected to be only the size of the data it holds?

Reply:
Hmm, not sure why this is an expectation from `DeviceBuffer`. I think we should get rid of it. Downstream things may break, but if that's the case, they should change their expectations too.

`sys.getsizeof` [should return the result of calling the `__sizeof__` method](https://docs.python.org/3.8/library/sys.html#sys.getsizeof) **plus some overhead**. This overhead is non-zero for objects that participate in garbage collection.

This [SO answer](https://stackoverflow.com/questions/42051378/sizeof-not-getting-called-by-sys-getsizeof) that I'm starting to dig into. But here's my understanding so far:

* `DeviceBuffer` was previously just a wrapper around `rmm::device_buffer` so it wasn't managed by Python's GC
* The addition of the `Stream` attribute (which itself holds a reference to Python object as its _owner_ attribute) means `DeviceBuffer` is now managed by Python's GC
* Thus, we can no longer expect `dbuf.__sizeof__() == sys.getsizeof(dbuf)`

---

As an aside, I'm not sure what the other implications are of `DeviceBuffer` now being handled by CPython's GC. @kkraus14 for visibility.

Reply:
That's good info @kkraus14 , what should we do moving forward? Are you saying should accept the fact that `sys.getsizeof` should return the result of calling the __sizeof__ method plus some overhead. FWIW, accepting that is my personal preference here, and I would then just fix the test.

Reply:
Yes, the test code is inherently broken where it should call `__sizeof__` directly if it wants to check that information.

Reply:
> > But if the stream was created outside of RMM there's no way it can be owned by RMM.
> 
> It doesn't need to be. The RMM Python library `Stream` class will hold a reference to a Python object which is what keeps the underlying `cudaStream_t` from being freed. We just need to keep a reference to a `Stream` instance to keep things alive.

It does need to be. If the stream was created outside of RMM, we don't know when the caller will call `cudaStreamDestroy` on the `cudaStream_t`. No matter how many references there are, it might be destroyed by its owner out from under you.  So it needs to be documented that doing that results in undefined behaviour.

Reply:
FYI, I just updated my comment above with a bit more accurate information for posterity.

Reply:
> It does need to be. If the stream was created outside of RMM, we don't know when the caller will call `cudaStreamDestroy` on the `cudaStream_t`. No matter how many references there are, it might be destroyed by its owner out from under you. So it needs to be documented that doing that results in undefined behaviour.

Documenting that makes total sense to me, I'll take care of that too. To expand a bit on the current cases we explicitly support.

1. Numba: Only calls `cuStreamDestroy` in [`_stream_finalizer`](https://github.com/numba/numba/blob/0bac18af44d08e913cd512babb9f9b7f6386d30a/numba/cuda/cudadrv/driver.py#L1343-L1347), which is registered as a [`Stream finalizer`](https://github.com/numba/numba/blob/0bac18af44d08e913cd512babb9f9b7f6386d30a/numba/cuda/cudadrv/driver.py#L1203-L1204)
1. CuPy: Only called in [`streamDestroy`](https://github.com/cupy/cupy/blob/b545c7bc67a7c9a2dc87c7a3a8b2ffeb2d45758c/cupy_backends/cuda/api/runtime.pyx#L786-L788), which is only called in [`Stream.__del__`](https://github.com/cupy/cupy/blob/b545c7bc67a7c9a2dc87c7a3a8b2ffeb2d45758c/cupy/cuda/stream.pyx#L320)

Of course, Python allow the user to do any crazy stuff, but from the Numba/CuPy design perspective I don't see anything to worry about.

Reply:
+1 to what @pentschev said. Most Python users are going to rely on Python's reference counting/GC to clean things up.

Reply:
I tried documenting this in https://github.com/rapidsai/rmm/pull/654/files#diff-f8cef9dac214b50e943de96a4eccf3d341f00b5187a9b54e63743e3f9f399ceaR54-R57 , and also fixed the test in https://github.com/rapidsai/rmm/pull/654/files#diff-edbfcbe7a4b6474859241cc360666a0d3962bffc01b6850c66ee0a75bc6850e0R113-R114 but I wasn't able to come up with anything more specific than that, if anybody has ideas please let me know.

I also added a test for the CuPy stream lifetime management in https://github.com/rapidsai/rmm/pull/654/files#diff-edbfcbe7a4b6474859241cc360666a0d3962bffc01b6850c66ee0a75bc6850e0R339-R351, which segfaults about 50% of the time without the stream reference in `DeviceBuffer` and doesn't anymore with that. I wasn't able to come up with a Numba test that segfaults, I can try a bit more next week but I'm not exactly sure whether this will work, otherwise the CuPy test probably covers the important part already.

Reply:
Thank you @pentschev . It's also possible that an app could have a stream that was created on the C++ side or in any other library, but I suppose that would be rare. In any case the same lifetime considerations apply.

Reply:
@harrism this should be addressed as the `Stream` Cython class has an `owner` attribute that holds a counted reference to a Python object which is expected to control the actual `cudaStream_t` lifetime.

Documentation was also updated.